lib/wallet/src/Cardano/Wallet/Write/Tx.hs

```diff
@@ -0,0 +1,481 @@
+        toStrict (Just a) = SJust a
+        toStrict Nothing = SNothing
+
+    downcastTxOut
+        :: TxOutInRecentEra
+        -> Either
+            ErrInvalidTxOutInEra
+            (Core.TxOut (ShelleyLedgerEra AlonzoEra))
+    downcastTxOut (TxOutInRecentEra _addr _val _datum (Just _script))
+        = Left ErrInlineScriptNotSupportedInAlonzo
+    downcastTxOut (TxOutInRecentEra _addr _val (Alonzo.Datum _) _script)
+        = Left ErrInlineDatumNotSupportedInAlonzo
+    downcastTxOut (TxOutInRecentEra addr val Alonzo.NoDatum Nothing)
+        = Right $ Alonzo.TxOut addr val SNothing
+    downcastTxOut (TxOutInRecentEra addr val (Alonzo.DatumHash dh) Nothing)
+        = Right $ Alonzo.TxOut addr val (SJust dh)
+
+--------------------------------------------------------------------------------
+-- UTxO
+--------------------------------------------------------------------------------
+
+-- | Construct a 'UTxO era' using 'TxIn's and 'TxOut's in said era.
+utxoFromTxOuts
+    :: RecentEra era
+    -> [(TxIn, Core.TxOut (ShelleyLedgerEra era))]
+    -> (Shelley.UTxO (ShelleyLedgerEra era))
+utxoFromTxOuts era = withStandardCryptoConstraint era $
+    Shelley.UTxO . Map.fromList
+
+-- | Construct a 'UTxO era' using 'TxOutInRecentEra'. Fails if any output is
+-- invalid in the targeted 'era'.
+utxoFromTxOutsInRecentEra
+    :: forall era. RecentEra era
+    -> [(TxIn, TxOutInRecentEra)]
+    -> Either ErrInvalidTxOutInEra (Shelley.UTxO (ShelleyLedgerEra era))
+utxoFromTxOutsInRecentEra era = withStandardCryptoConstraint era $
+    fmap (Shelley.UTxO . Map.fromList) . mapM downcast
+  where
+    downcast
+        :: (TxIn, TxOutInRecentEra)
+        -> Either
+            ErrInvalidTxOutInEra
+            (TxIn, TxOut (ShelleyLedgerEra era))
+    downcast (i, o) = do
+        o' <- unwrapTxOutInRecentEra era o
+        pure (i, o')
+
+--------------------------------------------------------------------------------
+-- Compatibility
+--------------------------------------------------------------------------------
+
+toCardanoUTxO
+    :: forall era. IsRecentEra era
+    => Shelley.UTxO (ShelleyLedgerEra era)
+    -> Cardano.UTxO era
+toCardanoUTxO = withConstraints $
+    Cardano.UTxO
+    . Map.mapKeys Cardano.fromShelleyTxIn
+    . Map.map (Cardano.fromShelleyTxOut (shelleyBasedEra @era))
+    . unUTxO
+  where
+    unUTxO (Shelley.UTxO m) = m
+
+    withConstraints
+        :: ((Crypto (Cardano.ShelleyLedgerEra era) ~ StandardCrypto) => a)
+        -> a
+    withConstraints a = case recentEra @era of
+        RecentEraBabbage -> a
+        RecentEraAlonzo -> a
+
+fromCardanoUTxO
+    :: forall era. IsRecentEra era
+    => Cardano.UTxO era
+    -> Shelley.UTxO (Cardano.ShelleyLedgerEra era)
+fromCardanoUTxO = withStandardCryptoConstraint (recentEra @era) $
+    Shelley.UTxO
+    . Map.mapKeys Cardano.toShelleyTxIn
+    . Map.map (Cardano.toShelleyTxOut (shelleyBasedEra @era))
+    . unCardanoUTxO
+  where
+    unCardanoUTxO (Cardano.UTxO m) = m
```

Comment:
Question: is there a round-trip test for `{from,to}CardandUTxO`?

Reply:
Added with the following note:

https://github.com/input-output-hk/cardano-wallet/blob/c525c1049f3b0984a27d68c11b523c478004adc5/lib/wallet/src/Cardano/Wallet/Write/Tx.hs#L472-L478